### DISCUSSION TITLE
How does the decoder with scorer always predict in vocab words?

### DISCUSSION THREAD

**Discussion by erksch** [coqui-ai/STT#2326] [Q&A]
Hey guys!

From experience with your models, I know that with an external scorer enabled (NGram LM) the model will never predict anything that is not in the vocabulary of the LM. But I was wondering how this actually works because I could not quite find that mechanism in the code.

`make_ngram` will convert a prefix into a list of words:

https://github.com/coqui-ai/STT/blob/bb75afb8556595cf152a27bd4382b649e59b8e83/native_client/ctcdecode/scorer.cpp#L369-L396

The scorer will give a harsh `OOV_SCORE` of `-1000.0` whenever a word is not in the language model vocab:

https://github.com/coqui-ai/STT/blob/bb75afb8556595cf152a27bd4382b649e59b8e83/native_client/ctcdecode/scorer.cpp#L329-L331 

But what happens if a character of a word is _silent_ and can not really be predicted by the acoustic model. 
This character would have a very low probability and is possibly pruned in the decoding step. 

If the character is missing, how can the NGram LM still reconstruct a full word and not always give OOV and discard the prefix? 

Or is it the job of the acoustic model to also output characters that are silent?

Coqui with LM scorer seems to behave like "find the closest valid hypothesis with full words in vocab" but how is this behaviour enforced?

**Comment by reuben** [comment]
You got it, it's not properly enforced, just heavily downscored. If you have lax enough pruning parameters then the decoder will still explore low prob AM labels which will then get boosted by LM scores of valid words.
